lib/bolt/config/transport.rb

```diff
@@ -0,0 +1,133 @@
+# frozen_string_literal: true
+
+require 'bolt/error'
+require 'bolt/util'
+
+module Bolt
+  class Config
+    class Transport
+      attr_reader :config
+
+      def initialize(data = {}, boltdir = nil)
+        assert_hash(data)
+        @config = Bolt::Util.deep_merge(defaults, filter(data))
```

Comment:
Updated to preserve the original input in `@input` and have `#merge` also accept objects of the same class. Merging will only merge `@input` and the data passed to `#merge` and then merge _that_ with the defaults and store it in `@config`.

I've left the`#to_h` method for now since I'm not sure how else we would retrieve the full config to pass to the applicator besides accessing individual options and building a hash with them e.g. `{ 'port' => config['port'], ... }`